compiler/vm/vmrunner.nim

```diff
@@ -0,0 +1,256 @@
+
+  c.allocator.byteType = c.typeInfoCache.charType
+
+  mapList(c.globals, p.globals, x):
+    c.heap.heapNew(c.allocator, c.types[x])
+
+  mapList(c.complexConsts, p.cconsts, x):
+    let
+      t = c.types[x.typ]
+      handle = c.allocator.allocConstantLocation(t)
+
+    discard loadConst(p, x.packedId.int, handle, c.memory)
+    handle
+
+  block:
+    # NOTE: callback handling is in a WIP state
+    # merges the already registered callbacks with the list stored in the
+    # packed state
+    let orig = c.callbacks
+    let origKeys = c.callbackKeys
+    c.callbacks.newSeq(p.callbacks.len)
+    c.callbackKeys.newSeq(p.callbacks.len)
+    for i, x in p.callbacks.pairs:
+      var overload: VmCallback
+      for j, y in origKeys.pairs:
+        if string(y) == x:
+          overload = orig[j]
+          break
+
+      if overload == nil:
+        # TODO: this should be an error
+        echo "no callback for key: ", x
+
+      c.callbackKeys[i] = IdentPattern(x)
+      c.callbacks[i] = overload
+
+proc loadFromFile(c: var TCtx, file: AbsoluteFile): Result[FunctionIndex, RodFileError] =
+  var p: PackedEnv
+  let err = readFromFile(p, file)
+  if err != RodFileError.ok:
+    result.initFailure(err)
+    return
+
+  loadIntoContext(c, p)
+  c.code = move p.code
+
+  result.initSuccess(p.entryPoint)
+
+
+proc main(args: seq[string]): int =
+  let config = newConfigRef(cli_reporter.reportHook)
+  config.writeHook =
+    proc(conf: ConfigRef, msg: string, flags: MsgFlags) =
+      msgs.msgWrite(conf, msg, flags)
+  config.writelnHook =
+    proc(conf: ConfigRef, msg: string, flags: MsgFlags) =
+      conf.writeHook(conf, msg & "\n", flags)
+
+  # setup the execution context
+  var c = TCtx(config: config)
+  c.features.incl(allowInfiniteLoops)
+
+  c.registerMathOps()
+  c.registerHashOps()
+
+  c.heap.slots.newSeq(1) # setup the heap
+
+  let lr = loadFromFile(c, toAbsolute(args[0], getCurrentDir().AbsoluteDir))
+  if lr.isErr:
+    # TODO: proper error reporting
+    echo lr.takeErr
+    return 1
+
+  let
+    entryPoint = c.functions[lr.unsafeGet.int]
+    cb = proc (c: TCtx, r: TFullReg): PNode =
+      c.config.internalAssert(r.kind == rkInt):
+        "expected int return value" # either the executable is malformed or
+                                    # there's an issue with the code-generator
+      newIntNode(nkIntLit, r.intVal)
+
+  # setup the starting frame:
+  var frame = TStackFrame(prc: entryPoint.sym, next: -1)
+  frame.slots.newSeq(entryPoint.regCount)
+
+  let r = c.execute(entryPoint.start, frame, cb)
+  if r.isOk:
+    # on successful execution, the executable's main function returns the
+    # value of ``programResult``, which we use as the runner's exit code
+    result = r.unsafeGet.intVal.int
+  else:
+    let err = r.takeErr
+    # an uncaught error occurred
+    c.config.localReport(err.stackTrace)
+    c.config.localReport(err.report)
+    result = 1
+
+programResult = main(commandLineParams())
```

Comment:
```suggestion
when defined(isMainModule):
  programResult = main(commandLineParams())
```

I think that should be fine and allow for direct module unit testing / inclusion?

Reply:
Done. `main` is now also exported .